src/udev/memory_id/dmidecode.c

```diff
@@ -0,0 +1,768 @@
+ *    Family "2.0", Level 00, Revision 00.43, January 26, 2015
+ *    https://trustedcomputinggroup.org/pc-client-platform-tpm-profile-ptp-specification/
+ *  - "RedFish Host Interface Specification" (DMTF DSP0270)
+ *    https://www.dmtf.org/sites/default/files/DSP0270_1.0.1.pdf
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <getopt.h>
+#define _GNU_SOURCE
+#include <errno.h>
+
+#include "build.h"
+#include "fileio.h"
+#include "udev-util.h"
+#include "types.h"
+
+#define out_of_spec "<OUT OF SPEC>"
+static const char *bad_index = "<BAD INDEX>";
+
+#define SUPPORTED_SMBIOS_VER 0x030300
+
+#define FLAG_NO_FILE_OFFSET     (1 << 0)
+#define FLAG_STOP_AT_EOT        (1 << 1)
+#define FLAG_FROM_DUMP          (1 << 2)
+static uint32_t opt_flags = 0;
+
+#define SYS_FIRMWARE_DIR "/sys/firmware/dmi/tables"
+#define SYS_ENTRY_FILE SYS_FIRMWARE_DIR "/smbios_entry_point"
+#define SYS_TABLE_FILE SYS_FIRMWARE_DIR "/DMI"
+
+static int checksum(const uint8_t *buf, size_t len) {
+        uint8_t sum = 0;
+        size_t a;
+
+        for (a = 0; a < len; a++)
+                sum += buf[a];
+        return (sum == 0);
+}
+
+/*
+ * Type-independant Stuff
+ */
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+        char *bp = (char *)dm->data;
+
+        bp += dm->length;
+        while (s > 1 && *bp)
+        {
+                bp += strlen(bp);
+                bp++;
+                s--;
+        }
+
+        if (!*bp)
+                return NULL;
+
+        return bp;
+}
+
+static const char *dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+        char *bp;
+
+        if (s == 0)
+                return "Not Specified";
+
+        bp = _dmi_string(dm, s);
+        if (bp == NULL)
+                return bad_index;
+
+        return bp;
+}
+
+/* shift is 0 if the value is in bytes, 1 if it is in kilobytes */
+static void dmi_print_memory_size(const char *attr_prefix, const char *attr_suffix, int slot_num, u64 code, int shift)
+{
+        uint64_t capacity;
+
+        capacity = code.h;
+        capacity = capacity << 32 | code.l;
+        if (shift == 1)
+                capacity = capacity << 10;
+
+        if (slot_num >= 0)
+                printf("%s_%u_%s=%"PRIu64"\n", attr_prefix, slot_num, attr_suffix, capacity);
+        else
+                printf("%s_%s=%"PRIu64"\n", attr_prefix, attr_suffix, capacity);
+}
+
+/*
+ * 7.17 Physical Memory Array (Type 16)
+ */
+
+static const char *dmi_memory_array_location(uint8_t code)
+{
+        /* 7.17.1 */
+        static const char *location[] = {
+                "Other", /* 0x01 */
+                "Unknown",
+                "System Board Or Motherboard",
+                "ISA Add-on Card",
+                "EISA Add-on Card",
+                "PCI Add-on Card",
+                "MCA Add-on Card",
+                "PCMCIA Add-on Card",
+                "Proprietary Add-on Card",
+                "NuBus" /* 0x0A */
```

Comment:
maybe use the `[…] = ` notation for arrays like this? i.e.

static const char *location[] = {
        [0x01] = "Other",
…
```

And yes, this would mean we'd waste the first entry, but I think that's OK.